File: src/main/java/bo/forja/backend/operacion/TiposDeclarados.java
package bo.forja.backend.operacion;

import java.util.Locale;
import java.util.Map;

/**
 * Tipos tal como los escribe o los dice una persona, llevados a un nombre
 * unico.
 * <p>
 * Lo comparten los canales que interpretan texto -el dictado y la lectura de
 * una fotografia- y esa union no es comodidad: si cada uno normalizara por su
 * cuenta, el mismo modelo terminaria con "texto" en un atributo y "String" en
 * otro, el generador emitiria dos tipos distintos para lo mismo y la
 * exportacion XMI declararia dos tipos de datos donde hay uno.
 * <p>
 * Un tipo que no esta en la tabla se respeta tal cual: puede ser otra clase del
 * modelo, o un enumerado que la persona va a crear despues. Inventar una
 * equivalencia seria peor que no tener ninguna.
 */
public final class TiposDeclarados {

    private static final Map<String, String> EQUIVALENCIAS = Map.ofEntries(
            // Texto
            Map.entry("texto", "String"), Map.entry("cadena", "String"),
            Map.entry("string", "String"), Map.entry("caracteres", "String"),
            Map.entry("char", "String"), Map.entry("varchar", "String"),
            Map.entry("text", "String"), Map.entry("str", "String"),
            // Enteros
            Map.entry("entero", "Integer"), Map.entry("numero", "Integer"),
            Map.entry("int", "Integer"), Map.entry("integer", "Integer"),
            Map.entry("number", "Integer"),
            Map.entry("largo", "Long"), Map.entry("long", "Long"),
            Map.entry("bigint", "Long"),
            // Decimales
            Map.entry("decimal", "Decimal"), Map.entry("importe", "Decimal"),
            Map.entry("precio", "Decimal"), Map.entry("monto", "Decimal"),
            Map.entry("real", "Decimal"), Map.entry("double", "Decimal"),
            Map.entry("float", "Decimal"), Map.entry("money", "Decimal"),
            // Booleanos
            Map.entry("booleano", "Boolean"), Map.entry("logico", "Boolean"),
            Map.entry("bool", "Boolean"), Map.entry("boolean", "Boolean"),
            // Fechas
            Map.entry("fecha", "Date"), Map.entry("date", "Date"),
            Map.entry("fechayhora", "DateTime"), Map.entry("fechahora", "DateTime"),
            Map.entry("datetime", "DateTime"), Map.entry("marcadetiempo", "DateTime"),
            Map.entry("timestamp", "DateTime"),
            Map.entry("hora", "Time"), Map.entry("time", "Time"),
            // Identificadores
            Map.entry("uuid", "UUID"), Map.entry("guid", "UUID"));

    private TiposDeclarados() {
    }

    public static String normalizar(String escrito) {
        if (escrito == null || escrito.isBlank()) {
            return "String";
        }
        String limpio = escrito.trim();
        String clave = limpio.toLowerCase(Locale.ROOT).replaceAll("[^\\p{L}\\p{N}]", "");

        String conocido = EQUIVALENCIAS.get(clave);
        if (conocido != null) {
            return conocido;
        }
        return Character.toUpperCase(limpio.charAt(0)) + limpio.substring(1);
    }
}

File: movil/lib/voz/tipos_declarados.dart
/// Tipos tal como los dice una persona, llevados a un nombre unico.
///
/// Espejo de TiposDeclarados.java. Del otro lado lo comparten el dictado y la
/// lectura de una fotografia, y esa union no es comodidad: si cada canal
/// normalizara por su cuenta, el mismo modelo terminaria con "texto" en un
/// atributo y "String" en otro, el generador emitiria dos tipos distintos para
/// lo mismo y la exportacion XMI declararia dos tipos de datos donde hay uno.
/// Por eso el telefono tiene que decir exactamente lo mismo.
///
/// Un tipo que no esta en la tabla se respeta tal cual: puede ser otra clase
/// del modelo, o un enumerado que la persona va a crear despues. Inventar una
/// equivalencia seria peor que no tener ninguna.

const _equivalencias = <String, String>{
  // Texto
  'texto': 'String', 'cadena': 'String', 'string': 'String',
  'caracteres': 'String', 'char': 'String', 'varchar': 'String',
  'text': 'String', 'str': 'String',
  // Enteros
  'entero': 'Integer', 'numero': 'Integer', 'int': 'Integer',
  'integer': 'Integer', 'number': 'Integer',
  'largo': 'Long', 'long': 'Long', 'bigint': 'Long',
  // Decimales
  'decimal': 'Decimal', 'importe': 'Decimal', 'precio': 'Decimal',
  'monto': 'Decimal', 'real': 'Decimal', 'double': 'Decimal',
  'float': 'Decimal', 'money': 'Decimal',
  // Booleanos
  'booleano': 'Boolean', 'logico': 'Boolean', 'bool': 'Boolean',
  'boolean': 'Boolean',
  // Fechas
  'fecha': 'Date', 'date': 'Date',
  'fechayhora': 'DateTime', 'fechahora': 'DateTime', 'datetime': 'DateTime',
  'marcadetiempo': 'DateTime', 'timestamp': 'DateTime',
  'hora': 'Time', 'time': 'Time',
  // Identificadores
  'uuid': 'UUID', 'guid': 'UUID',
};

String normalizarTipo(String? escrito) {
  if (escrito == null || escrito.trim().isEmpty) return 'String';

  final limpio = escrito.trim();
  final clave = limpio
      .toLowerCase()
      .replaceAll(RegExp(r'[^\p{L}\p{N}]', unicode: true), '');

  final conocido = _equivalencias[clave];
  if (conocido != null) return conocido;

  return limpio[0].toUpperCase() + limpio.substring(1);
}
